### PR TITLE
Fix (de)serialization of async search failures

### PR DESCRIPTION
**Pull request by jimczi**
The (de)serialization code of the async search response cannot handle exceptions that extend ElasticsearchException (e.g. ScriptException).
This commit fixes this bug by serializing the error with the more generic
StreamInput#writeException.
This change also adds the fetch failures that were missing in the response

### CODE REVIEW

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchResponseTests.java [hunk 1]
```diff
@@ -95,7 +96,7 @@ protected AsyncSearchResponse copyInstance(AsyncSearchResponse instance, Version
     }
 
     static AsyncSearchResponse randomAsyncSearchResponse(String searchId, SearchResponse searchResponse) {
-        int rand = randomIntBetween(0, 2);
+        int rand = randomIntBetween(2, 2);
```
Comment (jpountz) [inline x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchResponseTests.java:99]: leftover?

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java [hunk 2]
```diff
@@ -349,8 +349,10 @@ protected void onQueryFailure(int shardIndex, SearchShardTarget shardTarget, Exc
         }
 
         @Override
-        protected void onFetchFailure(int shardIndex, Exception exc) {
+        protected void onFetchFailure(int shardIndex, SearchShardTarget shardTarget, Exception exc) {
             checkCancellation();
+            searchResponse.get().addShardFailure(shardIndex,
+                new ShardSearchFailure(exc, shardTarget.getNodeId() != null ? shardTarget : null));
```
Comment (jpountz) [inline x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:355]: maybe add a comment explaining why we have this != null?

**Review by jpountz — APPROVED** [review]

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchResponseTests.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchResponseTests.java]: thanks

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:357]: I added https://github.com/elastic/elasticsearch/pull/55688/commits/245199625184a56b58d3e0524ccb55958b6ba494

**Pull request merged**